Clarify one sided payment rules

SWvheerden opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
RFC 201, for tari script, laid out a way to do one-sided payments. This RFC clarifies the calculation of the blinding factor k_b to ensure there is nothing left over to think about by explicitly typing out how to calculate it using Diffie-hellman. 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [ ] I'm merging against the `development` branch.
* [ ] I ran `cargo-fmt --all` before pushing.
* [ ] I ran `cargo test` successfully before submitting my PR.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

sdbondi commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -507,8 +506,10 @@ For Bob to claim his commitment, \\( C_b \\) he requires the blinding factor, \\
+who the sender is so that he can derive the shared secret.  
+
+This information can be obtained without needing to communicate with the sender. If the blinding factor \\( k_b \\) was calculated with diffie-hellman using the offset public keypair, (\\( k_{Ob} \\),\\( K_{Ob} \\)) as sender keypair and
+the keypair, (\\( k_{Sb} \\),\\( K_{Sb} \\)) as the receiver keypair, the blinding factor \\( k_b \\) can be securely calculated without communication between the 2 parties.
```
Not super necessary, but might be clearer to be 100% specific. 
i.e.
The sender calculates k_b = k_Ob * K_Sb, and the receiver calculates k_b = K_Ob * k_Sb

hansieodendaal replied:
Agreed

stringhandler replied:
How does the receiver know the value though?

hansieodendaal replied:
The value must be sent out of band.

stringhandler replied:
if the value is sent out of band, why not send the whole commitment out of band? Isn't the purpose here to not do out of band? 
"This information can be obtained with need to communicate with the sender"

sdbondi reviewed (dismissed):
LGTM

stringhandler reviewed (changes requested):
This doesn't seem to solve the problem of not communicating with the sender

hansieodendaal reviewed (commented):
Clarification of the Diffie-Hellman exchange is great.

**Discussion:**
Maybe we should add a short discussion about how the one-sided payment receiver can expect payments, but not exactly how much it would be, without communicating out of band with the sender. If the value of the payment is not known, the commitment cannot be opened without huge amounts of computing power.

One option is to know who will pay you and to provide them with a short identification string for one-sided payments, for example, an account number, that would be embedded as the proof message in the Bulletproof range proof. 
- Sender(s) knows
  - Receiver public wallet key
  - Tari one-sided payment script variants that would be accepted by the receiver
  - Unique sender proof message, e.g. an account number
- Receiver knows
  - Sender(s) public rewind key(s)
  - Sender(s) public wallet key(s)
  - Unique sender proof message

The receiver can then scan the blockchain for payments and open the commitments using the following method:
- Scan all UTXOs for all acceptable payment script hashes; there would only be a few at most, and if the script resolves when inserting the receiver public key, it is for the receiver.
- With those identified UTXOs, do Bulletproof rewinding with the range of registered sender public rewind keys. If the proof message matches, the rewound value is more than likely the correct value of the commitment.
- Test opening of the commitment:
  - Use the Diffie-Hellman exchange with corresponding public wallet key to derive the commitment's spending key and the value obtained from Bulletproof rewinding
  - If the commitment can be opened, it comes from the identified sender

stringhandler commented:
@hansieodendaal seems ok, but if the sender rewind keys are public, can't anyone just read all the values? Why not just include the value in metadata? Or maybe encrypt the value and then add it to the metadata?

hansieodendaal commented:
The sender's public rewind key is not common knowledge; it has to be shared out of band, and will only be shared with specific parties. The sender can also at any time change the rewind key or use more than one.

Although additional information can be added as metadata, we get up to 23 bytes worth of data for free that can be embedded in the Bulletproof rewind proof message. 

As you suggested, adding the value as encrypted metadata would also be feasible because the opening of the commitment can be tested with the decrypted value and all registered one-sided payment sender public keys. If the correct value was encrypted, the commitment can be opened.

SWvheerden commented:
Would using the blinding factor of the utxo or a derived key of that for the rewind key not solve the problem?

sdbondi commented:
I'm not very familiar with the RP rewinding maths but would it not be possible to add the value to the range proof using the Diffie-Hellman key `k_b`  or perhaps `Hash(k_b)` as the rewind blinding key? As part of scanning you would need to derive the DH key for each UTXO and attempt to rewind. 

This makes the problem of obtaining the value the same problem as "how do we efficiently/effectively scan for new one-sided payments?".

sdbondi commented:
@SWvheerden Beat me to it 😆

hansieodendaal commented:
> Would using the blinding factor of the utxo or a derived key of that for the rewind key not solve the problem?

This is a very good idea yes (@SWvheerden and @sdbondi) [using `Hash(k_b)`], as the sender will not want to share their own private rewind key (`pvt_rewind_key`). In this way, the sender could still use their own private blinding key (`pvt_blinding_key`) and keep it secret.
(See [Bulletproof rewinding](https://github.com/dalek-cryptography/bulletproofs/issues/335) for context)

SWvheerden commented:
> > ould using the blinding factor of the utxo or a derived key of that for the rewind key not solve the problem?

hansieodendaal commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -524,6 +524,10 @@ $$
 transaction to thwart any potential horizon attacks in the future.
 
+One other problem with one sided communication is the communication of the value of the transaction being sent, and the wallet recovery of one sided payments. We can solve both of these by using (\\( k_{Ob} \\).
+Because this is a secret key known by Alice and Bob and no one else. We hash this key as a new key, (\\( k_{rewind} =  Hash(k_{Ob}) \\). We know have a rewinding secret key for use in bulletproof rewinding. And now
```
We ~~know~~ _now_ have ...

stringhandler replied:
```suggestion
Because this is a secret key known by Alice and Bob and no one else. We hash this key as a new key, (\\( k_{rewind} =  Hash(k_b) \\). We know have a rewinding secret key for use in bulletproof rewinding. And now
```

hansieodendaal commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -524,6 +524,10 @@ $$
 
+One other problem with one sided communication is the communication of the value of the transaction being sent, and the wallet recovery of one sided payments. We can solve both of these by using (\\( k_{Ob} \\).
+Because this is a secret key known by Alice and Bob and no one else. We hash this key as a new key, (\\( k_{rewind} =  Hash(k_{Ob}) \\). We know have a rewinding secret key for use in bulletproof rewinding. And now
+we can encrypt the value (\\( v \\) inside of the bulletproof rewinding, allowing Bob to unlock his payment without revealing any information publicly.
```
~~we can encrypt the value (\\( v \\) inside of the bulletproof rewinding~~
we use the standard Bulletproof rewinding scheme to encrypt the value (\\( v \\) inside the Bulletproof range proof,

hansieodendaal commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -524,6 +524,10 @@ $$
 
+One other problem with one sided communication is the communication of the value of the transaction being sent, and the wallet recovery of one sided payments. We can solve both of these by using (\\( k_{Ob} \\).
+Because this is a secret key known by Alice and Bob and no one else. We hash this key as a new key, (\\( k_{rewind} =  Hash(k_{Ob}) \\). We know have a rewinding secret key for use in bulletproof rewinding. And now
+we can encrypt the value (\\( v \\) inside of the bulletproof rewinding, allowing Bob to unlock his payment without revealing any information publicly.
```
```suggestion
we use the standard Bulletproof rewinding scheme to encrypt the value (\( v \) inside the Bulletproof range proof,, allowing Bob to unlock his payment without revealing any information publicly. (See [Bulletproof rewinding](https://github.com/dalek-cryptography/bulletproofs/issues/335) for context)).
```

stringhandler replied:
Should be "We now use the standard...". Don't need to start a sentence with "And"

SWvheerden replied:
Do we have a link explaining it in our own repo, not Dalek's?

hansieodendaal replied:
We can always update it later when we have documented it ourselves.

hansieodendaal commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -524,6 +524,10 @@ $$
+One other problem with one sided communication is the communication of the value of the transaction being sent, and the wallet recovery of one sided payments. We can solve both of these by using (\\( k_{Ob} \\).
+Because this is a secret key known by Alice and Bob and no one else. We hash this key as a new key, (\\( k_{rewind} =  Hash(k_{Ob}) \\). We know have a rewinding secret key for use in bulletproof rewinding. And now
+we can encrypt the value (\\( v \\) inside of the bulletproof rewinding, allowing Bob to unlock his payment without revealing any information publicly.
+
```
I would add something like this:

-----------

In summary, receiving party scanning the blockchain for one-sided payments therefore involves:
- Scan all UTXOs trying to unlock the payment script hashes; if the script resolves when inserting the receiver public key, it is for the receiver.
- With those identified UTXOs, do Bulletproof rewinding with the range of registered sender public rewind keys ((\( k_{rewind}G \) as per above). If the proof message matches, the rewound value is more than likely the correct value of the commitment.
- Test opening of the commitment using the value obtained from rewinding and the spending key from the Diffie-Hellman exchange; if the commitment can be opened, it is intended for the receiver.

stringhandler replied:
Don't agree with point 2. Point 2 should be rewound using the hash of the blinding factor

hansieodendaal replied:
**Edit: ** _Strike that_ ~~We equate the private rewind key to the hash of the blinding factor. Bulletproof rewinding is done with the public rewind key, not the private rewind key.~~

Please see the explanation below.

stringhandler commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -507,12 +506,28 @@ For Bob to claim his commitment, \\( C_b \\) he requires the blinding factor, \\
 For Bob's part, when he discovers one-sided payments to himself, he should spend them to new outputs using a traditional
 transaction to thwart any potential horizon attacks in the future.
 
+One other problem with one sided communication is the communication of the value of the transaction being sent, and the wallet recovery of one sided payments. We can solve both of these by using (\\( k_{Ob} \\).
```
```suggestion
One other problem with one sided communication is the communication of the value of the transaction being sent, and the wallet recovery of one sided payments. We can solve both of these by using (\\( k_b \\).
```

SWvheerden replied:
We should be using k_{Ob}, the offset pubkey. k_{b} is the blinding factor.

stringhandler replied:
Bob does not know `k_{Ob}` though?

SWvheerden replied:
Yes, He only knows `K_{Ob}` and `k_S`

stringhandler replied:
The wording is ambiguous, but the next line is correct, so will let it pass

stringhandler commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -482,11 +482,23 @@ Once again, transaction fees are ignored to simplify the illustration.
+
+Bob requires the value \\( v_b \\) and blinding factor \\( k_b \\) to claim his payment, but he needs to be able to claim it without asking Alice for them.
+
+This information can be obtained by using diffie-hellman and bulletproof rewinding. If the blinding factor \\( k_b \\) was calculated with diffie-hellman using the offset public keypair, (\\( k_{Ob} \\),\\( K_{Ob} \\)) as sender keypair and
```
```suggestion
This information can be obtained by using Diffie-Hellman and bulletproof rewinding. If the blinding factor \\( k_b \\) was calculated with diffie-hellman using the offset public keypair, (\\( k_{Ob} \\),\\( K_{Ob} \\)) as sender keypair and
```

stringhandler commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -482,11 +482,23 @@ Once again, transaction fees are ignored to simplify the illustration.
+    k_b = k_{Ob} * {K_Sb}
+$$
+
+Next alice next uses bulletproof rewinding to encrypt the value \\( v_b \\) into the the bulletproof for the commitment \\( C_b \\). For this she uses (\\( k_{rewind} =  Hash(k_{Ob}) \\) as the encryption key.
```
```suggestion
Next Alice next uses bulletproof rewinding to encrypt the value \\( v_b \\) into the the bulletproof for the commitment \\( C_b \\). For this she uses (\\( k_{rewind} =  Hash(k_{Ob}) \\) as the encryption key.
```

stringhandler commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -495,20 +507,27 @@ $$
+    k_b = K_{Ob} * {k_Sb}
+$$
+
+Next Bob's wallet calculates \\( k_{rewind} \\), using \\( k_{rewind} =  Hash(k_{Ob})\\) and uses that to rewind the bulletproof to get the value \\( v_b \\). 
```
Where does Bob find out `k_{0}b` in order to hash it?

stringhandler reviewed (commented):
Seems like this refactor has reverted to some previous errors

hansieodendaal commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -495,20 +507,27 @@ $$
+    k_b = K_{Ob} * {k_Sb}
+$$
+
+Next Bob's wallet calculates \\( k_{rewind} \\), using \\( k_{rewind} =  Hash(k_{b})\\) and uses that to rewind the Bulletproof to get the value \\( v_b \\). 
```
**Edit:** _Strike that_ ~~We equate the private rewind key to the hash of the blinding factor. Bulletproof rewinding is done with the public rewind key, not the private rewind key.~~

hansieodendaal replied:
We actually need to convey two values, not only one. To be clear on the previous comment, Bulletproof rewinding is done with the two public rewind keys, not the private rewind keys.

From [Tari fork of Bulletproofs]( https://github.com/tari-project/bulletproofs/blob/a18a3d2c91cea4e5f999446a566a4b0af7c9e187/tests/range_proof.rs#L154)
``` rust
    // The two rewind keys can be shared with a trusted 3rd party, whom will be able to extract
    // the value of the commitment as well as the extra data
    let pub_rewind_key_1 =
        RistrettoPoint::from(&pvt_rewind_key * &RISTRETTO_BASEPOINT_TABLE).compress();
    let pub_rewind_key_2 =
        RistrettoPoint::from(&pvt_blinding_key * &RISTRETTO_BASEPOINT_TABLE).compress();
    // The rewind nonces are derived from the public rewind keys and the Pedersen commitment
    let rewind_nonce_1 = get_rewind_nonce_from_pub_key(&pub_rewind_key_1, &committed_value);
    let rewind_nonce_2 = get_rewind_nonce_from_pub_key(&pub_rewind_key_2, &committed_value);
    // Get secret nonces from the private keys - this emulates the owner of a wallet
    let blinding_nonce_1 = get_secret_nonce_from_pvt_key(&pvt_rewind_key, &committed_value);
    let blinding_nonce_2 = get_secret_nonce_from_pvt_key(&pvt_blinding_key, &committed_value);

    // Get Value Test 1 - provide correct rewind nonces
    let mut rewind_transcript = Transcript::new(b"Bulletproof-Rewind Test");
    assert_eq!(
        proof.rewind_single_get_value_only(
            &bp_gens,
            &mut rewind_transcript,
            &committed_value,
            64,
            &rewind_nonce_1,
            &rewind_nonce_2,
        ),
        Ok((confidential_value, proof_message))
    );
```

hansieodendaal commented on RFC/src/RFC-0201_TariScript.md:
```diff
@@ -482,11 +482,23 @@ Once again, transaction fees are ignored to simplify the illustration.
+    k_b = k_{Ob} * {K_Sb}
+$$
+
+Next Alice next uses Bulletproof rewinding to encrypt the value \\( v_b \\) into the the Bulletproof for the commitment \\( C_b \\). For this she uses (\\( k_{rewind} =  Hash(k_{b}) \\) as the encryption key.
```
We need two private keys here: `pvt_rewind_key` and `pvt_blinding_key`. This could be something like:
```
k_{rewind} =  Hash(k_{b})
k_{blinding} =  Hash(Hash(k_{b}) | k_{b}))
```
Please see the Bulletproofs code snippet for the full explanation.

SWvheerden committed:
clarify one sided payment

add explicit formula

Add value encryption

refactor

Update RFC/src/RFC-0201_TariScript.md

Co-authored-by: Mike the Tike <stringhandler@gmail.com>
Update RFC/src/RFC-0201_TariScript.md

Co-authored-by: Mike the Tike <stringhandler@gmail.com>
fix error

update bulletproof to Bulletproof

edit

SWvheerden committed:
add note

hansieodendaal reviewed (approved):
LGTM
